Support defining remote components in Go

justinvp opened the pull request:
This makes it possible to create a resource provider with a `Construct` implementation in Go, similar to [what we have for Node](https://github.com/pulumi/pulumi/tree/master/sdk/nodejs/provider).

Fixes #5489

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+
+// ProviderMain is the typical entrypoint for a resource provider plugin. Using it isn't required but can cut down
+// significantly on the amount of boilerplate necessary to fire up a new resource provider.
+func ProviderMain(name, version string, args ProviderArgs) error {
```
I'd originally wanted to put this in it's own `provider` package and name this function `Main` -- but then I wouldn't have been able to access some of the private details in `pulumi` that are needed for `Construct`'s implementation. Definitely open to alternative names for this.

lukehoban replied:
It is a little unfortunate that this all ends up in the top-level package - especially since some concepts overlap a bit with existing concepts with the same or similar names (like `Provider`).  Do we have an option to expose a smaller new surface area publicly here and then the rest of the API in a child package?  Not a blocker though if this is sufficiently unnatural.

justinvp replied:
I'm going to noodle on this a bit this week to see what we can do.

justinvp replied:
Two things: 1) The APIs are now exposed in the `pulumi/provider` package, rather than the top-level `pulumi` package and 2) the API has been whittled down to no longer expose a provider API. Instead, there is a `Construct` function that adapts the gRPC request/response to/from the Pulumi Go programming model, to be used in conjunction with the existing provider helper we have in https://github.com/pulumi/pulumi/tree/master/pkg/resource/provider.

I've also added a convenience function `ComponentMain` in the `pkg/resource/provider` package, that can be used to implement a minimal resource provider just for components, so you don't have to write all the other resource provider boilerplate just for `Construct`. I have a draft PR of the go-boilerplate repo that demonstrates its use: https://github.com/pulumi/pulumi-component-provider-go-boilerplate/pull/3

```go
import (
	"github.com/pulumi/pulumi/pkg/v3/resource/provider"
	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
)

// Serve launches the gRPC server for the resource provider.
func Serve(providerName, version string, schema []byte) {
	// Start gRPC service.
	if err := provider.ComponentMain(providerName, version, schema, construct); err != nil {
		cmdutil.ExitError(err.Error())
	}
}
```

```go
import (
	"github.com/pkg/errors"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/provider"
)

func construct(ctx *pulumi.Context, typ, name string, inputs provider.ConstructInputs,
	options pulumi.ResourceOption) (*provider.ConstructResult, error) {
	// TODO: Add support for additional component resources here.
	switch typ {
	case "xyz:index:StaticPage":
		return constructStaticPage(ctx, name, inputs, options)
	default:
		return nil, errors.Errorf("unknown resource type %s", typ)
	}
}

// constructStaticPage is an implementation of Construct for the example StaticPage component.
// It demonstrates converting the raw ConstructInputs to the component's args struct, creating
// the component, and returning its URN and state (outputs).
func constructStaticPage(ctx *pulumi.Context, name string, inputs provider.ConstructInputs,
	options pulumi.ResourceOption) (*provider.ConstructResult, error) {

	// Copy the raw inputs to StaticPageArgs. `inputs.SetArgs` uses the types and `pulumi:` tags
	// on the struct's fields to convert the raw values to the appropriate Input types.
	args := &StaticPageArgs{}
	if err := inputs.SetArgs(args); err != nil {
		return nil, errors.Wrap(err, "setting args")
	}

	// Create the component resource.
	staticPage, err := NewStaticPage(ctx, name, args, options)
	if err != nil {
		return nil, errors.Wrap(err, "creating component")
	}

	// Return the component resource's URN and state. `NewConstructResult` automatically sets the
	// ConstructResult's state based on resource struct fields tagged with `pulumi:` tags with a value
	// that is convertible to `pulumi.Input`.
	return provider.NewConstructResult(staticPage)
}
```

I've also included a `NewConstructResult` helper along the lines of your suggestion, that creates the result with the URN and state based on the outputs defined on the component.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+
+// ProviderMain is the typical entrypoint for a resource provider plugin. Using it isn't required but can cut down
+// significantly on the amount of boilerplate necessary to fire up a new resource provider.
+func ProviderMain(name, version string, args ProviderArgs) error {
```
I considered providing something that could work with this existing helper in pkg https://github.com/pulumi/pulumi/blob/master/pkg/resource/provider/main.go (possibly exposing a func from the Go sdk `pulumi` package that could be passed to `pkg/v2/resource/provider.Main`, but that felt a bit cumbersome, so went with a new "main" func here in `sdk` instead.

justinvp replied:
I ended up reworking this so you can use it with the existing https://github.com/pulumi/pulumi/blob/master/pkg/resource/provider. I added a convenience `ComponentMain` there if you only care about a resource provider for components, but if you wanted to have a resource provider that had both custom and component resources that's now possible.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+
+// ProviderArgs allows specifying implementations of the Provider interface for orchestrating resource create, read,
+// update, and delete operations.
+type ProviderArgs struct {
```
I'd originally wanted to name this `Provider`, but we already have a `Provider` identifier in this package (a function for specifying the `Provider` resource option).

justinvp replied:
No longer applicable.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+	// Schema is the schema for the provider in JSON.
+	Schema []byte
+
+	// Check validates that the given property bag is valid for a resource of the given type and returns the inputs
+	// that should be passed to successive calls to Diff, Create, or Update for this resource.
+	Check func(ctx *ProviderContext, urn resource.URN, olds,
+		news resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error)
```
I decided to pass a new `*ProviderContext` type to each of the regular CRUD operations, to provide a means of sending log messages to the engine. We didn't do that for the Node.js API, though. Curious what folks think about this. It's sort of analogous to the `*HostClient` that is handed to the callback in `pkg/v2/resource/provider.Main`.

https://github.com/pulumi/pulumi/blob/277dee6cc14d2b6d3c795edd3bc1010b246e0f5f/pkg/resource/provider/main.go#L36

https://github.com/pulumi/pulumi/blob/277dee6cc14d2b6d3c795edd3bc1010b246e0f5f/pkg/resource/provider/host.go#L35-L95

justinvp replied:
No longer applicable.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+}
+
+// SetArgs sets the inputs on the given args struct.
+func (inputs *ConstructInputs) SetArgs(args interface{}) error {
```
The `inputs` arg to the user-provided `Construct` func is typed as `*ConstructInputs`. Originally, I was going to type this argument as `pulumi.Map` (`map[string]Input`) and pass the inputs like:

```go
	inputs := Map{}
 	for k, v := range deserializedInputs {
 		output := newOutput(anyOutputType, v.deps...)
 		output.resolve(v.value, true /*known*/, v.secret, nil)
 		inputs[k] = output
 	}
```

However, this wasn't all that useful in practice. For example, one of the components I was working on had a `tags` input that was typed as `map[string]string` (though, after deserializing it was actually a `map[string]interface{}`). Inside my provided `Construct` implementation I was having to manually convert the `AnyOutput` to `StringMapOutput` via an `ApplyT`, which wasn't a great experience (having to do that for each property).

So, instead I created a `ConstructInputs` type with two methods:

- `Map()` returns the inputs as a `pulumi.Map` with all the inputs as `AnyOutput`.
- `SetArgs(args interface{})` is a convenience helper. Pass it a pointer to an input struct, and it will use reflection to determine the appropriate output type.

An example of using SetArgs:

```go
type VpcArgs struct {
	// Requests an Amazon-provided IPv6 CIDR block with a /56 prefix length for the VPC. You cannot
	// specify the range of IP addresses, or the size of the CIDR block. Default is `false`.  If set
	// to `true`, then subnets created will default to `assignIpv6AddressOnCreation: true` as well.
	AssignGeneratedIpv6CidrBlock pulumi.BoolInput `pulumi:"assignGeneratedIpv6CidrBlock"`

	// The CIDR block for the VPC. Defaults to "10.0.0.0/16" if unspecified.
	CidrBlock pulumi.StringInput `pulumi:"cidrBlock"`

	// A boolean flag to enable/disable ClassicLink
	// for the VPC. Only valid in regions and accounts that support EC2 Classic.
	// See the [ClassicLink documentation][1] for more information. Defaults false.
	EnableClassiclink pulumi.BoolInput `pulumi:"enableClassiclink"`

	// A boolean flag to enable/disable ClassicLink DNS Support for the VPC.
	// Only valid in regions and accounts that support EC2 Classic.
	EnableClassiclinkDnsSupport pulumi.BoolInput `pulumi:"enableClassiclinkDnsSupport"`

	// A boolean flag to enable/disable DNS hostnames in the VPC. Defaults to true if unspecified.
	EnableDnsHostnames pulumi.BoolInput `pulumi:"enableDnsHostnames"`

	// A boolean flag to enable/disable DNS support in the VPC. Defaults true if unspecified.
	EnableDnsSupport pulumi.BoolInput `pulumi:"enableDnsSupport"`

	// A tenancy option for instances launched into the VPC. Defaults to "default" if unspecified.
	InstanceTenancy pulumi.StringInput `pulumi:"instanceTenancy"` // TODO enum "default" | "dedicated"?

	// A mapping of tags to assign to the resource.
	Tags pulumi.StringMapInput `pulumi:"tags"`
}

// In my Construct for this component:
args := &VpcArgs{}
err := inputs.SetArgs(args)
if err != nil {
    return pulumi.ConstructResult{}, errors.Wrap(err, "setting args")
}
```

lukehoban replied:
> Does having the SetArgs helper make sense? Should we do this in some other way?

FWIW - As a consumer of this - the `SetArgs` helper feels very useful to project args into a typed input Args bag for a component.  

In fact, it made me want something similar in reverse - so that this code could be replaced with `return pulumi.ConstructOutputsFromComponent(staticPage)`:

```go
	// Return the component resource's URN and outputs as its state.
	return pulumi.ConstructResult{
		URN: staticPage.URN(),
		State: pulumi.Map{
			"bucket":     staticPage.Bucket,
			"websiteUrl": staticPage.WebsiteUrl,
		},
	}, nil
```

Doing so would reduce the number of places I have to touch when updating the public API surface area of a component from 3 to 2 (just the schema and implementation).

> I didn't want to force component authors to have to define it twice manually like this. But curious if you think we should force component authors to define these the same way as the Go codegen does.

I believe the only place the `vpcArgs` args is needed is for enabling `VpcArgs` to be an `Input` itself (with an underlying value type of `vpcArgs`).  I don't think there is any need to require component authors of Go-based multi-language components to implement that.

Would love @pgavlin or @EvanBoyle or @viveklak  confirmation on this from a Go component model perspective.

justinvp replied:
> In fact, it made me want something similar in reverse - so that this code could be replaced with `return pulumi.ConstructOutputsFromComponent(staticPage)`:

Yes! I've wanted the exact same helper. I added `SetArgs` out of necessity, but would also love something along the lines of your suggestion. I'll add it to this PR sometime before next week when I have a spare moment.

Aside: Ultimately, we're going to want to significantly improve the experience here such that someone doesn't even have to write `Construct` at all for components. They'll just write their Go component as normal, and perhaps call some registration function, passing it the type of the args and resource, and then a schema will be able to be automatically generated from that as well as the construct implementation, so the user doesn't have to deal with any of this, and doesn't have to deal with keeping a separate schema.json up-to-date.

justinvp replied:
> I believe the only place the `vpcArgs` args is needed is for enabling `VpcArgs` to be an `Input` itself (with an underlying value type of `vpcArgs`). I don't think there is any need to require component authors of Go-based multi-language components to implement that.

Ah, that makes sense.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+// representation of the properties as present in the program inputs. Though this rule is not
+// required for correctness, violations thereof can negatively impact the end-user experience, as
+// the provider inputs are using for detecting and rendering diffs.
+func (p *providerServer) Check(ctx context.Context, req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
```
Note: The implementation of these is based on the implementations in https://github.com/pulumi/pulumi/blob/34eae7e7e3ddae475c52f2e248b1eabb23d62cf7/sdk/go/common/resource/plugin/provider_server.go

justinvp replied:
No longer applicable.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+	Delete func(ctx *ProviderContext, urn resource.URN, id resource.ID, props resource.PropertyMap,
+		timeout float64) error
+
+	// Construct creates a new component resource.
+	Construct func(ctx *Context, typ, name string, inputs *ConstructInputs,
+		options ResourceOption) (ConstructResult, error)
```
I didn't want to force authors to have to implement an entire provider interface just to support `Construct`, so I've made it so you pass these as fields on this struct.

justinvp replied:
No longer applicable.

lukehoban commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,792 @@
+
+		val, secret, err := unmarshalPropertyValue(pulumiCtx, input)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unmarshaling input %s", k)
```
Trying to use https://github.com/pulumi/pulumi-component-provider-go-boilerplate.  I'm seeing that my example program just silently doesn't create the component I specified (`const api = new apigateway.RestAPI("api", {...})` is in my code, and runs, but neither it nor any of the child resources are created, and no error is raised).  

When I look at verbose logs, I see this:


> I0326 14:54:57.706385   26028 eventsink.go:59] RegisterResource RPC finished: resource:api[apigateway:index:RestAPI]; err: Error: 2 UNKNOWN: unmarshaling input routes: missing one or more required arguments, resp: undefined

But then after that the registration still seems to complete "correctly" just without doing anything.  

It appears that the error raised on this line isn't causing the resource operation to fail.  I'm unclear if that's an issue on this side of the RPC contract, or within the engine itself.

justinvp replied:
I believe Evan's Node.js error handling fix has addressed the error not failing the update. The actual issue you were running into was around the lack of prompt value support, which is on the way.

justinvp commented:
I've force pushed a new commit that addresses the feedback and adds integration tests. I am finishing support for prompt values for a separate PR, rather than including it here.

Here's a summary of what's changed:

- The public API is now exported from the `sdk/v3/go/pulumi/provider` package rather than the top-level `sdk/v3/go/pulumi` package. I decided to use `go:linkname` to make a small number of unexported functions implemented inside `pulumi` available inside the `pulumi/provider` package. This way, we can access the unexported functionality in `pulumi` but expose the public API from `pulumi/provider`. This is consistent with what the Go `runtime` package does to expose some functions implemented inside the `runtime` package to other packages.

    - As an alternative, I'd considered adding an `pulumi/internal` package and moving a bunch of functionality from `pulumi` into it and exporting it, so that both `pulumi` and `pulumi/provider` could access it, but this would have been a more significant refactoring.

    - If the use of `go:linkname` is undesirable (I'm having a hard time coming up with a major downside, other than it being a little unusual and having to maintain the package version in the `//go:linkname` comments), we could consider moving `Construct` into the top-level `pulumi` package.

- The `pulumi/provider` package only expose a relatively small surface area, basically a `Construct` adapter function and that converts the gRPC request/response to/from the Pulumi Go programming model. I've removed the provider implementation that was included. Instead, the `Construct` adapter is intended to be used in conjunction with the _existing_ helper APIs we have in https://github.com/pulumi/pulumi/tree/master/pkg/resource/provider. And to make it even easier for components, I've added a new `ComponentMain` function there, which implements a resource provider with the minimal amount of support for components, so you don't have to write a boilerplate provider implementation yourself.

- I've added a `func NewConstructResult(resource pulumi.ComponentResource) (*ConstructResult, error)` helper to avoid the boilerplate around return the component's URN and state.

- I've opened a draft PR of the go boilerplate repo that updates it to use the changes from this PR: https://github.com/pulumi/pulumi-component-provider-go-boilerplate/pull/3

lukehoban commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,271 @@
+				continue
+			}
+
+			if !field.Type.Implements(reflect.TypeOf((*Input)(nil)).Elem()) {
```
[I just found these two comments in drafts that I had intended to send a couple weeks ago!  I believe they are still relevant to the design here though - so sharing now ahead of a more thorough review of the updated PR shortly]

> I didn't want to force component authors to have to define it twice manually like this. But curious if you think we should force component authors to define these the same way as the Go codegen does.

Two things that I've gotten tripped up on myself working with this:
1. How do I write my own nested type to use here?  Like if I want a `Routes: []Route`, how do I write `[]Route` so that it works?  It appears I need to create a `type RouteArrayInput`, ensure that has a `ToRouteArrayOutputWithContext`, and then also a custom `Output` struct that wraps the raw `Route` struct?  I haven't yet quite figured out what works fully here.
2. If we had just alllowed using the same thing that codegen produces, then I could just use the types that the Go SDK codegen already generates for me - by having the provider actually depend on the SDK.  That may be a terrible idea, but it is nice in that the implementation gets to work with the exact same structs/interfaces as consumers will.  It also means I don't have to recreate all of this manually inside the implementation - since there is already a codegen process to create these for the SDK.   However, the types implemented in the SDK don't seem to work correctly with the logic as implemented here.

lukehoban replied:
It's also a little strange that these accept Input types - given that at least currently they will always be Outputs, and even once prompt values are possible, the user will always want to be clear here in the types which ones are prompt and which are Output. It will never be desirable for the implementor to see Input types vs. Output types. The implementation will always have to immediately turn the inputs back into outputs to work with them.

If this args struct used Outputs, the implementation here would also be much simpler and less "weird" in that it wouldn't have to discover To<>OutputWithContext to figure out what the output type was to create.

justinvp replied:
> How do I write my own nested type to use here?

The idea was to follow our documented pattern for inputs:

https://github.com/pulumi/pulumi/blob/afe109621e878aaf717044bf9c7f0958c53f6ffa/sdk/go/pulumi/types.go#L767-L829

But it is an unfortunate amount of boilerplate to have to write manually (especially so if you'd need to also manually write associated boilerplate for array variants, map variants, etc.), so I fully empathize with the desire from (2) to use the types that the Go SDK codegen generates. However, ideally, we'd have a way to write a much simpler type in the implementation without all the extra ceremony/boilerplate, and drive the schema + SDK generation based on it, not the other way around.

> However, the types implemented in the SDK don't seem to work correctly with the logic as implemented here.

Hmmm, do you happen to recall the specific issue? I believe there is a bug in the current implementation in this PR, where the unmarshalled value from the gRPC call isn't converted into the struct type -- so it'll remain a map. I am addressing this issue with the changes for prompt values.

> It's also a little strange that these accept Input types - given that at least currently they will always be Outputs, and even once prompt values are possible, the user will always want to be clear here in the types which ones are prompt and which are Output. It will never be desirable for the implementor to see Input types vs. Output types. The implementation will always have to immediately turn the inputs back into outputs to work with them.
>
> If this args struct used Outputs, the implementation here would also be much simpler and less "weird" in that it wouldn't have to discover To<>OutputWithContext to figure out what the output type was to create.

This is a good point. I was thinking you'd write the component implementation the same way you'd write it if you were releasing it as a regular Go library, where (ignoring prompt values for a moment) you would want to type inputs as the `Input` type to support either `Input` or `Output` values (and inside the implementation you'd immediately convert to `Output` to work with it).

But for the implementation of multi-lang components in Go, I could see how typing it as `Output` would simplify things here. We'd immediately know the `Output` type to use and it would save the user from having to convert the value to `Output` inside the component implementation.

I do still feel like we could support `Input` though... I could see supporting either `Output` or `Input`. If it's `Output`, great -- use it. If it's `Input`, one thing I could do is get its `ElementType()` and then see if the output type has been registered with `pulumi.RegisterOutputType` (using it if so), and finally falling back to looking for a `To<>Output` method.

> discover To<>OutputWithContext

Note: after the recent discussion of possibly deprecating the `<>WithContext` methods, I've switched this to look for just `To<>Output`.

justinvp commented on sdk/go/pulumi/provider.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+// constructInputsSetArgs sets the inputs on the given args struct.
+func constructInputsSetArgs(inputs map[string]interface{}, args interface{}) error {
```
Note: the implementation of this function isn't quite right, so it's not super important to review the details of it for this PR. I've reimplemented it for prompt values (separate PR on the way), and as part of that, it's leveraging `rpc.unmarshalOutput` to unmarshal the value.

pgavlin commented on pkg/resource/provider/component_provider.go:
```diff
@@ -0,0 +1,162 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provider
```
Any reason not to use [`plugin.providerServer`](https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/plugin/provider_server.go)? Looks like their functionality is similar.

justinvp replied:
The `Construct` adapter func in this PR adapts the raw gRPC interface to/from the Pulumi Go SDK model.

https://github.com/pulumi/pulumi/blob/a8ea363df970c58709097b3ab28e82afc5a0d9ce/sdk/proto/go/provider.pb.go#L2047-L2048

So the provider in this file implements the raw gRPC interface.

To use `plugin.providerServer`, `Construct` would have to be an adapter of `plugin.providerServer`'s interface:

https://github.com/pulumi/pulumi/blob/664776efe85ca4c6a8c9cd95fa2fbfa75f2dbf03/sdk/go/common/resource/plugin/provider.go#L78-L80

Should `Construct` be an adapter for `plugin.providerServer` instead? Or should we have adapters for both the raw gRPC interface **and** `plugin.providerServer`?

The reason I made `Construct` an adapter for the raw gRPC interface is because 1) our existing [provider boilerplate repo](https://github.com/pulumi/pulumi-provider-boilerplate) implements a provider using the raw gRPC interface rather than `plugin.providerServer` and 2) our own native providers also implement the raw gRPC interface over `plugin.providerServer`.

pgavlin replied:
> Should Construct be an adapter for plugin.providerServer instead? Or should we have adapters for both the raw gRPC interface and plugin.providerServer?
> 
> The reason I made Construct an adapter for the raw gRPC interface is because 1) our existing provider boilerplate repo implements a provider using the raw gRPC interface rather than plugin.providerServer and 2) our own native providers also implement the raw gRPC interface over plugin.providerServer.

IMO we should adapt from `plugin.providerServer`, but I think that we can move forward with what you have. In the fullness of time, I'd love for all of our Go providers to be sitting on top of `providerServer`.

pgavlin commented on sdk/go/pulumi/provider/empty.s:
```diff
@@ -0,0 +1,17 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This empty .s file enables the use of go:linkname to make certain unexported functions from
```
Interesting--are you sure this file is necessary? I think I've used `//go:linkname` w/o something like this in the past.

justinvp replied:
Without it there are compilation errors about missing function bodies.

pgavlin replied:
Weird. SGTM, then 👍
